LibPCAP support for Amazon Dash Binding

It seems at this point that the Amazon Dash binding requires a separate install and permission settings to accommodate detecting Dash buttons. This requires install of libpcap drivers to sniff the wire. I know some folks may not configure Docker to allow this host to be able to sniff the wire, but if setup with a macvlan network or other customer configs, the adapter should be able to sniff the wire properly. 

I would like to see the extensibility of adding this setup to support the Dash binding to the core docker container, so that if I do this manually it won't be overwritten anytime an update is pushed to the image. Also in the case of testing with other setups, it would be useful to not need to reinstall. 

Reference Binding Installation: http://docs.openhab.org/addons/bindings/amazondashbutton/readme.html#prerequisites

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/42377300-libpcap-support-for-amazon-dash-binding?utm_campaign=plugin&utm_content=tracker%2F31868624&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F31868624&utm_medium=issues&utm_source=github).
</bountysource-plugin>

Hi,

i also tried to get the Dash Button working with the docker image on the raspberry3 but it does not work.

The container is completely fresh and with libpcap support using the documentation and docker-compose

The error message i got is:

`2017-03-26 23:16:01.250 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occured while initializing handler of thing 'amazondashbutton:dashbutton:6e23f5f5': java.lang.NoClassDefFoundError: Could not initialize class com.sun.jna.Pointer`

Hi,

i have the same issue on my OpenHAB instance. I checked out the sources and played around with them a bit, trying to figure out the source of the error.
The first thing i noticed is, that no network interfaces have been found. So i digged deeper and added a logger to the catch-block of PcapUtil's getAllNetworkInterfaces method, which gave me the following:

        2017-09-18 12:54:08.166 [ERROR] [zondashbutton.internal.pcap.PcapUtil] - Cannot retrieve NetworkInterfaces
        org.pcap4j.core.PcapNativeException: Return code: -1, Message: Can't open netlink socket 13:Permission denied
                at org.pcap4j.core.Pcaps.findAllDevs(Pcaps.java:58)[256:org.openhab.binding.amazondashbutton:2.2.0.201709181253]
                at org.openhab.binding.amazondashbutton.internal.pcap.PcapUtil.getAllNetworkInterfaces(PcapUtil.java:34)[256:org.openhab.binding.amazondashbutton:2.2.0.201709181253]
                at org.openhab.binding.amazondashbutton.internal.pcap.PcapNetworkInterfaceService.determineBoundNetworkInterfaces(PcapNetworkInterfaceService.java:181)[256:org.openhab.binding.amazondashbutton:2.2.0.201709181253]
                at org.openhab.binding.amazondashbutton.internal.pcap.PcapNetworkInterfaceService.access$2(PcapNetworkInterfaceService.java:179)[256:org.openhab.binding.amazondashbutton:2.2.0.201709181253]
                at org.openhab.binding.amazondashbutton.internal.pcap.PcapNetworkInterfaceService$1.run(PcapNetworkInterfaceService.java:62)[256:org.openhab.binding.amazondashbutton:2.2.0.201709181253]
                at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_144]
                at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)[:1.8.0_144]
                at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)[:1.8.0_144]
                at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)[:1.8.0_144]
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)[:1.8.0_144]
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)[:1.8.0_144]
                at java.lang.Thread.run(Thread.java:748)[:1.8.0_144]

To me, it seems like some flag has been forgotten to set or docker is corrupt. I'm using docker 1.13.1 and used a slightly modified version of the docker-compose.yml that includes NET_ADMIN and NET_RAW, offered at dockerhub.

Does someone know how to fix this?

--- a/README.md
+++ b/README.md
@@ -195,9 +195,8 @@ volumes:
 
 ### Running openHAB with libpcap support
 
-You can run all openHAB images with libpcap support.
-This enables you to use the *Amazon Dashbutton Binding* in the Docker container.
-For that feature to work correctly, you need to run the image as **root user**.
+You can run all openHAB images with libpcap support by configuring additional capabilities.
+This allows for using the *Amazon Dash Button Binding* in the Docker container.
 Create the following `docker-compose.yml` and start the container with `docker compose up -d`
 
 ```yaml
@@ -218,12 +217,7 @@ services:
       - "./openhab_addons:/openhab/addons"
       - "./openhab_conf:/openhab/conf"
       - "./openhab_userdata:/openhab/userdata"
-    # The command node is very important. It overrides
-    # the "gosu openhab tini -s ./start.sh" command from Dockerfile and runs as root!
-    command: "tini -s ./start.sh server"
 ```
-
-*If you could provide a method to run libpcap support in user mode please open a pull request.*
 
 ### Running on Windows and macOS
 

--- a/alpine/entrypoint
+++ b/alpine/entrypoint
@@ -10,6 +10,19 @@ if [ "${CRYPTO_POLICY}" = "unlimited" ]; then
   sed -i 's/^crypto.policy=limited/crypto.policy=unlimited/' "${JAVA_HOME}/conf/security/java.security"
 fi
 
+# Set capabilities when available for add-ons using Pcap4J
+if capsh --print | grep -E 'Current:.+,cap_net_admin,cap_net_raw,.+' >/dev/null; then
+  setcap cap_net_raw,cap_net_admin=eip "${JAVA_HOME}/bin/java"
+  cat >> "/etc/ld-musl-$(arch).path" <<-EOI
+/lib
+/usr/lib
+${JAVA_HOME}/lib
+${JAVA_HOME}/lib/jli
+${JAVA_HOME}/lib/security
+${JAVA_HOME}/lib/server
+EOI
+fi
+
 # Deleting instance.properties to avoid karaf PID conflict on restart
 # See: https://github.com/openhab/openhab-docker/issues/99
 rm -f "${OPENHAB_HOME}/runtime/instances/instance.properties"

--- a/debian/entrypoint
+++ b/debian/entrypoint
@@ -10,6 +10,11 @@ if [ "${CRYPTO_POLICY}" = "unlimited" ]; then
   sed -i 's/^crypto.policy=limited/crypto.policy=unlimited/' "${JAVA_HOME}/conf/security/java.security"
 fi
 
+# Set capabilities when available for add-ons using Pcap4J
+if capsh --print | grep -E 'Current:.+,cap_net_admin,cap_net_raw,.+' >/dev/null; then
+  setcap cap_net_raw,cap_net_admin=eip "${JAVA_HOME}/bin/java"
+fi
+
 # Deleting instance.properties to avoid karaf PID conflict on restart
 # See: https://github.com/openhab/openhab-docker/issues/99
 rm -f "${OPENHAB_HOME}/runtime/instances/instance.properties"

